File: 0208-implement-trie-prefix-tree/0208-implement-trie-prefix-tree.cpp
class TrieNode{
    public:
        bool end;
        TrieNode* child[26];
        TrieNode(){
            for(int i = 0;i < 26;i++){
                child[i] = NULL;
            }
            end = false;
        }
        void setChild(char ch){
            int idx = (ch-'a');
            child[idx] = new TrieNode();
        }
        TrieNode* getChild(char ch){
            int idx = (ch-'a');
            return child[idx];
        }
        void setEnd(){
            end = true;
        }
        bool getEnd(){
            return end;
        }
};
class Trie {
public:
    TrieNode* root;
    Trie() {
        root = new TrieNode();
    }
    
    void insert(string word) {
        TrieNode* node = root;
        for(char ch: word){
            if(node->getChild(ch)==NULL){
                node->setChild(ch);
            }
            node = node->getChild(ch);
        }
        node->setEnd();
    }
    
    bool search(string word) {
        TrieNode* node = root;
        for(char ch: word){
            if(node->getChild(ch)==NULL){
                return false;
            }
            node = node->getChild(ch);
        }
        return node->getEnd();
    }
    
    bool startsWith(string prefix) {
        TrieNode* node = root;
        for(char ch: prefix){
            if(node->getChild(ch)==NULL){
                return false;
            }
            node = node->getChild(ch);
        }
        return true;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */

File: 0208-implement-trie-prefix-tree/0208-implement-trie-prefix-tree.java
class TrieNode{
    TrieNode child[];
    boolean end;
    TrieNode(){
        child = new TrieNode[26];
        end = false;
    }
    boolean getEnd(){
        return end;
    }
    void setEnd(){
        end = true;
    }
    TrieNode getChild(char ch){
        return child[ch-'a'];
    }
    void setChild(char ch){
        child[ch-'a'] = new TrieNode();
    }
}
class Trie {
    TrieNode root;
    public Trie() {
        root = new TrieNode();
    }
    
    public void insert(String word) {
        TrieNode node = root;
        int n = word.length();
        // for(char ch: word){
        // }
        // node.setEnd();
        for(int i = 0;i < n;i++){
            char ch = word.charAt(i);
            if(node.getChild(ch)==null){
                node.setChild(ch);
            }
            node = node.getChild(ch);
        }
        node.setEnd();
    }
    
    public boolean search(String word) {
        TrieNode node = root;
        int n = word.length();
        for(int i = 0;i < n;i++){
            char ch = word.charAt(i);
            if(node.getChild(ch)==null){
                return false;
            }
            node = node.getChild(ch);
        }
        return node.getEnd();
    }
    
    public boolean startsWith(String prefix) {
        TrieNode node = root;
        int n = prefix.length();
        for(int i = 0;i < n;i++){
            char ch = prefix.charAt(i);
            if(node.getChild(ch)==null){
                return false;
            }
            else{
                node = node.getChild(ch);
            }
        }
        return true;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */